cpp/core/shuffle/ShuffleWriter.cc

```diff
@@ -124,6 +126,17 @@ int64_t ShuffleBufferPool::bytesAllocated() const {
   return pool_->bytes_allocated();
 }
 
+ShuffleWriter::ShuffleWriter(
+    int32_t numPartitions,
+    std::shared_ptr<PartitionWriterCreator> partitionWriterCreator,
+    const ShuffleWriterOptions& options)
+    : numPartitions_(numPartitions),
+      options_(options),
+      pool_(std::make_shared<ShuffleBufferPool>(options_.memory_pool)),
+      codec_(createArrowIpcCodec(options_.compression_type, options_.codec_backend)) {
+  partitionWriter_ = partitionWriterCreator->make(this).ValueOrDie();
```

Comment:
Let's avoid directly calling ValueOrDie here. Please refer to Arrow's doc https://arrow.apache.org/docs/cpp/api/support.html#_CPPv4NKR5arrow6Result10ValueOrDieEv

Reply:
Let me explain why I made this change.

I think partitionWriterCreator is used to create partition writer, and this creating action can be done in the constructor of ShuffleWriter, The partitionWriterCreator's task is done after that and we don't need it anymore.

We need not keep partitionWriterCreator as the member data of VeloxShuffleWriter and be used in the init function of VeloxShuffleWriter.

Reply:
I agree with your point. But what I concern is that we should avoid calling `ValueOrDie` here.

Reply:
Maybe we can do more refactors here. Currently we have `LocalPartitionWriterCreator` and `CelebornPartitionWriterCreator`. Should it be better to create one factory for making the `PartitionWriter` instance?

Reply:
The XxCreator is taking the role of object factory now.

Reply:
using `ARROW_ASSIGN_OR_RAISE` here will cause returning errors of compiling, so if you think it's a problem, let me revert this change here.

Reply:
I revert this change, let's keep the old style.